syntax = "proto3";

package greet;

option go_package="./greetpb";

option java_package = "com.mesadhan.greet";
option java_multiple_files = true;

message Greeting {
    string first_name = 1;
    string last_name = 2;
}

// unary
message GreetRequest {
    Greeting greeting = 1;
}
message GreetResponse {
    string result = 1;
}



// server streaming
message GreetManyTimesRequest {
    Greeting greeting = 1;
}
message GreetManyTimesResponse {
    string result = 1;
}



// client streaming
message LongGreetRequest {
    Greeting greeting = 1;
}
message LongGreetResponse {
    string result = 1;
}



// BI-DI streaming
message GreetEveryoneRequest {
    Greeting greeting = 1;
}
message GreetEveryoneResponse {
    string result = 1;
}

service GreetService{

    rpc Greet(GreetRequest) returns (GreetResponse) {}    // Unary
    rpc GreetManyTimesMethod (GreetManyTimesRequest) returns (stream GreetManyTimesResponse) {};    // Server Streaming
    rpc LongGreet (stream LongGreetRequest) returns (LongGreetResponse) {};     // client streaming
    rpc GreetEveryoneMethod (stream GreetEveryoneRequest) returns (stream GreetEveryoneResponse) {};    // Bidirectional Streaming

}